fs_mark/fs_mark_clean.c

```diff
@@ -0,0 +1,99 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Cleanup after fs_mark (executed with files kept).
+ *
+ * Copyright 2023 Phoenix Systems
+ * Author: Adam Debek
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <stdint.h>
+#include <errno.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <stdio.h>
+#include <dirent.h>
+#include <string.h>
+#include <limits.h>
+
+int remove_dir_recursive(const char *dirPath)
+{
+	DIR *dir;
+	char entryPath[PATH_MAX];
+	struct dirent *entry;
+	int (*remove_func)(const char *);
+	int ret, try = 0;
+
+	dir = opendir(dirPath);
+	if (dir == NULL) {
+		fprintf(stderr, "Opendir() failed errno: %s\n", strerror(errno));
+		return -1;
+	}
+
+	if (dir) {
+		while ((entry = readdir(dir)) != NULL) {
+			if (!strcmp(".", entry->d_name) || !strcmp("..", entry->d_name)) {
+				continue;
+			}
+			sprintf(entryPath, "%s/%s", dirPath, entry->d_name);
+			remove_func = entry->d_type == DT_DIR ? remove_dir_recursive : remove;
+			if (remove_func(entryPath) != 0) {
+				closedir(dir);
+				return -1;
+			}
+		}
+
+		if (closedir(dir)) {
+			return -1;
+		}
+	}
+
+	errno = 0;
+	while ((ret = rmdir(dirPath)) < 0) {
+		if (errno == ENOTEMPTY && try < 5) {
+			remove_dir_recursive(dirPath);
+			try++;
+		}
+		else if (errno == ENOENT) {
+			errno = 0;
+			return 0;
+		}
+		else {
+			return -1;
+		}
```

Comment:
Is it necessary? We can remove dirs also using `remove()`.

Reply:
But using `rmdir()` is semantically correct.